diff --git a/asdf.lisp b/asdf.lisp
index 24f1172..57e015a 100644
--- a/asdf.lisp
+++ b/asdf.lisp
@@ -826,17 +826,35 @@ actually-existing directory."
 (defmethod component-relative-pathname ((component module))
   (let ((specified-pathname (or (slot-value component 'relative-pathname)
                                 (component-name component))))
-    (if (pathnamep specified-pathname)
-        specified-pathname
-        (multiple-value-bind (relative path)
-            (component-name-to-pathname-components specified-pathname t)
-          (make-pathname
-           :directory `(,relative ,@path)
-           :host (pathname-host (component-parent-pathname component)))))))
+    (etypecase specified-pathname
+      (pathname specified-pathname)
+      ((or symbol string)
+       (let ((*default-pathname-defaults*
+              (make-pathname :name nil :type nil :directory nil :host nil
+                             :defaults (component-pathname (component-parent component))))
+             (coerced-pathname (pathname (string specified-pathname))))
+         (etypecase coerced-pathname
+           (logical-pathname coerced-pathname)          ; must be absolute
+           (pathname
+            (make-pathname :directory `(:relative
+                                        ,@(rest (pathname-directory coerced-pathname))
+                                        ,@(when (pathname-name coerced-pathname)
+                                            (list (pathname-name coerced-pathname))))))))))))
 
 (defmethod component-pathname ((component component))
-  (merge-pathnames (component-relative-pathname component)
-                   (component-parent-pathname component)))
+  (let ((crp (component-relative-pathname component))
+        (cpp (component-parent-pathname component)))
+    (ecase (first (pathname-directory crp))
+      ((nil)
+       (make-pathname :name (pathname-name crp) :type (pathname-type crp)
+                      :defaults cpp))
+      (:absolute
+       crp)
+      (:relative
+       (make-pathname :directory (append (pathname-directory cpp)
+                                         (rest (pathname-directory crp)))
+                      :name (pathname-name crp) :type (pathname-type crp)
+                      :defaults cpp)))))
 
 (defmethod component-property ((c component) property)
   (cdr (assoc property (slot-value c 'properties) :test #'equal)))
